### PR TITLE
[ci] Add scheduled push to nightly feed job

### PR DESCRIPTION
**Pull request by pjcollins**
Imports the steps that push packages to the nightly feed from the
[release definition][0]. This package push will now happen as part of a
scheduled run of the MAUI-Release build pipeline.

[0]: https://dev.azure.com/devdiv/DevDiv/_release?_a=releases&view=mine&definitionId=2847

### CODE REVIEW

File: eng/pipelines/maui-release.yml [hunk 1]
```diff
@@ -190,4 +196,85 @@ extends:
...
+                inputs:
+                  azureSubscription: maui-nightly-feed-mi
+                  scriptType: ps
+                  scriptLocation: inlineScript
+                  inlineScript: |
+                    $accessToken = az account get-access-token --query accessToken --resource $(MauiNightlyFeedMIResourceId) -o tsv
+                    Write-Host "##vso[task.setsecret]$accessToken"
+                    Write-Host "Overwriting authorization token for the maui-nightly-feed service connection"
+                    # https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#setendpoint-modify-a-service-connection-field
+                    Write-Host "##vso[task.setendpoint id=$(MauiNightlyFeedServiceConnectionId);field=authParameter;key=apitoken]${accessToken}"
```
Comment (pjcollins) [inline eng/pipelines/maui-release.yml:229]: Nightly variables are set in the build definition: https://dev.azure.com/devdiv/DevDiv/_apps/hub/ms.vss-ciworkflow.build-ci-hub?_a=edit-build-definition&id=15722&view=Tab_Variables

File: eng/pipelines/maui-release.yml [hunk 2]
```diff
@@ -190,4 +196,85 @@ extends:
...
+                  $dropPrefix = $jsonContent.VstsDropBuildArtifact.VstsDropUrl -replace 'https://devdiv.artifacts.visualstudio.com/DefaultCollection/_apis/drop/drops/' -replace '/nugets'
+                  Write-Host "##vso[task.setvariable variable=ReleaseDropPrefix]$dropPrefix"
+                displayName: Set variable ReleaseDropPrefix
+
+              - task: ms-vscs-artifact.build-tasks.artifactDropDownloadTask-1.artifactDropDownloadTask@1
+                displayName: Download $(ReleaseDropPrefix)/nugets
+                inputs:
+                  dropServiceURI: https://devdiv.artifacts.visualstudio.com/DefaultCollection
+                  buildNumber: $(ReleaseDropPrefix)/nugets
+                  destinationPath: $(Build.StagingDirectory)\nuget-signed
```
Comment (pjcollins) [inline eng/pipelines/maui-release.yml:253]: This currently relies on https://github.com/dotnet/maui/pull/23658, but could be updated to download from `nuget-signed` and `vs-msi-nugets` artifacts instead for other branches.

Comment (pjcollins) [inline eng/pipelines/maui-release.yml:253]: This currently relies on https://github.com/dotnet/maui/pull/23658, but should probably be updated to download from `nuget-signed` and `vs-msi-nugets` artifacts instead, assuming we only want to use this to .NET 8 builds from main for now.

```suggestion
              - task: DownloadPipelineArtifact@2
                inputs:
                  artifactName: nuget-signed
                  downloadPath: $(Build.StagingDirectory)\nuget-signed
                displayName: Download nuget-signed

             - task: DownloadPipelineArtifact@2
                inputs:
                  artifactName: vs-msi-nugets
                  downloadPath: $(Build.StagingDirectory)\nuget-signed
                displayName: Download vs-msi-nugets
```

File: eng/pipelines/maui-release.yml [hunk 3]
```diff
@@ -190,4 +196,85 @@ extends:
...
+                  scriptType: ps
+                  scriptLocation: inlineScript
+                  inlineScript: |
+                    $accessToken = az account get-access-token --query accessToken --resource $(MauiNightlyFeedMIResourceId) -o tsv
+                    Write-Host "##vso[task.setsecret]$accessToken"
+                    Write-Host "Overwriting authorization token for the maui-nightly-feed service connection"
+                    # https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#setendpoint-modify-a-service-connection-field
+                    Write-Host "##vso[task.setendpoint id=$(MauiNightlyFeedServiceConnectionId);field=authParameter;key=apitoken]${accessToken}"
+
+              - task: NuGetAuthenticate@0
```
Comment (mjbond-msft) [inline eng/pipelines/maui-release.yml:231]: Use latest `NuGetAuthenticate@1` instead

File: eng/pipelines/maui-release.yml [hunk 4]
```diff
@@ -190,4 +196,85 @@ extends:
...
+                              & dotnet nuget push --source $feed --api-key az --skip-duplicate $nupkgFile
+                              $waiting = $false
+                          }
+                          catch {
+                              if ($attempt -gt $maxAttempts)
+                              {
+                                throw 'Maximum attempts reached, failing!'
+                              }
+                              Write-Output "  attempt $attempt of $maxAttempts failed..."
+                              $attempt = $attempt + 1
```
Comment (mjbond-msft) [inline eng/pipelines/maui-release.yml:276]: How about a sleep here with incremental backoff?

Comment (rmarinho) [reply eng/pipelines/maui-release.yml:276]: I think this is just a copy of what we have in the release, we can improve, but this seems to be working for us till now :)

File: eng/pipelines/maui-release.yml (hunk 3 above)
Comment (pjcollins) [reply eng/pipelines/maui-release.yml:231]:
```suggestion
              - task: NuGetAuthenticate@1
```

**Comment by pjcollins** [comment]
Seems like this may be ready to merge and iterate as needed, though probably best to try one test run against a manually queued build first?

**Comment by rmarinho** [comment]
> Seems like this may be ready to merge and iterate as needed, though probably best to try one test run against a manually queued build first?

Seems is giving some error 

```

Encountered error(s) while parsing pipeline YAML:
/eng/pipelines/maui-release.yml (Line: 199, Col: 11): Unexpected symbol: ')'. Located at position 42 within expression: eq(variables['Build.Reason'], 'Schedule')). For more help, refer to https://go.microsoft.com/fwlink/?linkid=842996
```

File: eng/pipelines/maui-release.yml [hunk 5]
```diff
@@ -190,4 +196,78 @@ extends:
             os: ${{ parameters.VM_IMAGE_HOST.os }}     
             scanArtifacts: ['${{ parameters.PackPlatform.binariesArtifact }}']
 
-        
+        - ${{ if eq(variables['Build.Reason'], 'Schedule')) }}:
```
Comment (pjcollins) [inline eng/pipelines/maui-release.yml:199]:
```suggestion
        - ${{ if eq(variables['Build.Reason'], 'Schedule') }}:
```

**Comment by pjcollins** [comment]
> > Seems like this may be ready to merge and iterate as needed, though probably best to try one test run against a manually queued build first?
> 
> Seems is giving some error
> 
> ```
> 
> Encountered error(s) while parsing pipeline YAML:
> /eng/pipelines/maui-release.yml (Line: 199, Col: 11): Unexpected symbol: ')'. Located at position 42 within expression: eq(variables['Build.Reason'], 'Schedule')). For more help, refer to https://go.microsoft.com/fwlink/?linkid=842996
> ```

I should have fixed this, but you will probably need to change this condition to `- ${{ if eq(variables['Build.Reason'], 'Manual') }}:` to test it

**Review by rmarinho — APPROVED** [review]

**Pull request merged**